### PR TITLE
Feat(cli): added the offline flag to the dry-run command

### LINKED ISSUE #3865: [Feature] Add a "template" command to the vela cli
**Is your feature request related to a problem? Please describe.**
Its important to template out the k8s resources in the CI Pipeline to run tools like Datree for policy management. Currently there is a `vela dry-run` but it requires access to a k8s cluster (kube api server to be more precise). This isn't possible in most ci layers.

**Describe the solution you'd like**
A `vela template` command to spit out the k8s resources without requiring a kubeapi. We can provide it a directory containing all the trait and component definitions so it doesn't have to query kubernetes.

**Describe alternatives you've considered**
Dry run the resources and commit it in git alongside the application file. This isn't the best option.

**Additional context**

### PR DESCRIPTION
**Pull request by YourTechBud**
### Description of your changes

Fixes #3865

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [x] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Made a sample `definition.yaml` file and a `app.yaml` and run the `vela dry-run` command with the `-d` `-f` and `--offline` flags.

### CODE REVIEW

File: pkg/appfile/template.go [hunk 1]
```diff
@@ -297,7 +297,12 @@ func DryRunTemplateLoader(defs []oam.Object) TemplateLoaderFn {
 			}
 		}
 		// not found in provided cap definitions
-		// then try to retrieve from cluster
+		// then try to retrieve from cluster if not in offline mode where reader will be nil
+		if r == nil {
+			return nil, fmt.Errorf("cannot load template %q from provided ones", capName)
```
Comment (wonderflow) [inline pkg/appfile/template.go:302]:
```suggestion
			return nil, fmt.Errorf("cannot load template %q from provided ones in offline mod", capName)
```

**Review by wonderflow — COMMENTED** [review]
`discoverymapper.DiscoveryMapper`

What will the discoverymapper.DiscoveryMapper work in offline mode?

**Comment by YourTechBud** [comment]
> `discoverymapper.DiscoveryMapper`
> 
> What will the discoverymapper.DiscoveryMapper work in offline mode?

During my testing the discovery mapper wasnt invoked at all if all the components were found in the definition file. Let me check it out once to understand its utility.

**Comment by YourTechBud** [comment]
> What will the discoverymapper.DiscoveryMapper work in offline mode?

@wonderflow I have replaced `client.Client` with a fake client for offline mode. This will help us retain the behavior of `ExecuteDryRun` as long as the provided definitions contain all the dependencies. This will take care of `discoverymapper.DiscoveryMapper` and the parser (`parsePolicies`, `parseWorkflowSteps` and `parseReferredObjects`) which earlier depended upon an actual client

**Review by wonderflow — APPROVED** [review]
LGTM, thanks

**Comment by YourTechBud** [comment]
@wonderflow are any more changes required in this PR?

**Comment by wonderflow** [comment]
Ping @Somefive , Please review

**Review by Somefive — APPROVED** [review]

**Pull request merged**